src/lsm/forest.zig

```diff
@@ -514,47 +540,24 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
             }
 
             // On the last beat of the bar, make sure that manifest log compaction is finished.
-            if (last_beat or last_half_beat) {
-                switch (forest.manifest_log_progress) {
-                    .idle => {},
-                    .compacting => unreachable,
-                    .done => {
-                        forest.manifest_log.compact_end();
-                        forest.manifest_log_progress = .idle;
-                    },
-                    .skip => {},
-                }
+            if ((last_beat or last_half_beat) and op > constants.lsm_compaction_ops) {
+                forest.manifest_log.compact_end();
             }
 
-            const callback = progress.callback;
+            const callback = forest.progress.compact.callback;
             forest.progress = null;
 
             callback(forest);
         }
 
-        fn compact_manifest_log_callback(manifest_log: *ManifestLog) void {
-            const forest: *Forest = @fieldParentPtr("manifest_log", manifest_log);
-            assert(forest.manifest_log_progress == .compacting);
-
-            forest.manifest_log_progress = .done;
-
-            if (forest.progress) |progress| {
-                assert(progress == .compact);
-
-                forest.compact_callback();
-            } else {
-                // The manifest log compaction completed between compaction beats.
```

Comment:
This branch is now an assert. VOPRing that specifically at https://github.com/tigerbeetle/tigerbeetle/pull/2290